### ISSUE TITLE
Provide guidance for workflows broken by @4?

### ISSUE THREAD

**Issue by nedbat** [actions/upload-artifact#472]
### What files would you like to change?

I can see that @4 brought improvements in speed and reliability, that's great.  But it also broke a number of common patterns in workflows.  Can you provide some guidance about ways to achieve the same outcomes with the new version?  This might involved new features in actions/download-artifact.

### What are your suggested changes?

I'm looking for help, I don't have solutions yet myself.

**Comment by henryiii** [comment]
Personally, I think two changes would really help: If there was some way for download-artifact to extract multiple artifacts into a single directory, and if there was some way to use a glob to express the artifacts to download. They could even be together, such as specifying `coverage-*` as the name extracting all matching files into one directory, like one name does. Or it could be a toggle.

FYI, this affects basically every cibuildwheel user (there are over a thousand https://github.com/pypa/cibuildwheel/issues/1699), as it was the way you'd merge binaries on different platforms before uploading to PyPI. It also affects the recommendations from the PyPI and Scientific-Python Development Guidelines, etc.

FYI, people are merging the v4 updates without realizing the breakage, as Dependabot is making them separately, so if they appear broken, they still go with it assuming it's the v3/v4 incompatibility, and if it's only used for the release step (common with cibuildwheel workflows), then they won't even know it's broken until they try to release.

Happy to update as much as I can to whatever the recommendation is. My current plan is:

```yaml
    - uses: actions/upload-artifact@v4
      with:
        name: Wheel-${{ matrix.<whatever> }}
...

    - uses: actions/download-artifact@v4
      with:
        path: all

    - name: Merge files
      run: |
        mkdir dist
        mv all/Wheel-*/* .
```

But besides the extra step, will also download everything, even if there's stuff like coverage files that are not needed for that job.

**Comment by webknjaz** [comment]
Even in the context wider than just uploading to PyPI, when I build a bunch of wheels on different platforms, there should be a way to collect them all for testing (examples: https://github.com/aio-libs/yarl/blob/88900a1/.github/workflows/ci-cd.yml#L208-L265 /https://github.com/aio-libs/frozenlist/blob/b5ffcd7/.github/workflows/ci-cd.yml#L207-L264)

**Comment by webknjaz** [comment]
This also breaks the use of composite actions that themselves make use of either upload-artifacts or download-artifacts:
- https://twitter.com/hynek/status/1735568463093997752
- https://hynek.me/articles/ditch-codecov-python/
- https://github.com/hynek/build-and-inspect-python-package/releases/tag/v2.0.0

I'm curious if the change is an attempt to reduce occurrences of the race condition I reported a while back: https://github.com/actions/toolkit/issues/1235 / https://github.com/actions/download-artifact/issues/140#issuecomment-1314062872.

**Comment by woodruffw** [comment]
As an additional datapoint here: GitHub's own docs for publishing to PyPI via OIDC encourage the "upload-then-download" pattern, which (I believe) will be broken by these action changes when used with multiple uploaders (which is common in Python, for builds on different hosts or platforms).

(This has the unfortunate potential to break a large percentage of Python publishing workflows.)

Ref: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-pypi

**Comment by per1234** [comment]
>If there was some way for download-artifact to extract multiple artifacts into a single directory

From my experiments while trying to find a way to adapt the hundreds of workflows I maintain that rely on the ability to upload multiple times from matrix jobs to the same artifact for transfer to a dependent job that consumes the content of the artifact, I find the `actions/download-artifact@v4` action already has that capability.

For example, a workflow job with these steps will succeed and show that the files from the `foo` and `bar` artifacts are present under the `baz` folder as expected.

```yaml
- uses: actions/download-artifact@v4
  with:
    name: foo
    path: baz

- uses: actions/download-artifact@v4
  with:
    name: bar
    path: baz

- run: ls baz
```

If you found some conditions where the action is not able to do this, please do submit a report to the issue tracker for the **actions/download-artifact** as I agree that it will be a common use pattern for those of us who are migrating from using a single artifact to multiple artifacts for transferring files from one job to another.

> if there was some way to use a glob to express the artifacts to download

I agree. This request is already tracked at https://github.com/actions/download-artifact/issues/6

**Comment by henryiii** [comment]
The problem with repeating the action over and over is that for a normal matrix for coverage or wheel builds, there might now be dozens of artifacts (one per Python version × OS × arch × PyPy/CPython, for example). Also, if one or more jobs fail, you still might want to upload coverage, which means you'd also need to do something like `always()`. Last I checked I think cibuildwheel could produce about 70+ different wheels if you targeted everything it can handle. You can also dynamically generate the matrix, see https://iscinumpy.dev/post/cibuildwheel-2-10-0/#only, which is how MyPyC does it. A normal coverage set is about 20 or so (5 Python versions + 3 PyPy versions × 3 OSs).

**Comment by woodruffw** [comment]
Another example of a pattern that will now fail here: workflows that upload coverage from multiple jobs for later combination now get `409 Conflict`.

One example of such a workflow: https://github.com/sigstore/sigstore-python/blob/13f513761473888f32df942246055b0567264223/.github/actions/upload-coverage/action.yml

And example workflow failure: https://github.com/sigstore/sigstore-python/actions/runs/7234398921/job/19710726122?pr=841

**Comment by webknjaz** [comment]
_(my attempt to summarize the important pain point contexts)_

@per1234 your example features a linear workflow with a known static artifact names, but this issue refers to a different practice of CI configuration which involves matrixes.

Note all other people who commented here, combined maintain hundreds of workflows too. What unites us is that we all come from the Python ecosystem and rely on the ability to upload a number of artifacts from jobs in matrixes of varying sizes and different factor naming strategies. We essentially need a way to combine many files coming from different jobs (not necessarily one file per job, though — but usually, all uploaded in one go) in a "synchronization point" — a job that waits for that matrix and does something with those files.

Two major contexts in which this is happening are:
1. Publishing a bunch of Python distribution packages to PyPI — this has to happen together for all files, once we know all the previous jobs are complete.
2. Combining coverage data produced in separate jobs into one, for determining if the overall metrics meet the configured threshold.

With different artifact names, we'd need to loop over the matrix artifacts somehow to get all those files together. And GitHub Actions workflows don't have syntax for such loops. Moreover, this means every project reinventing a wheel of naming the workflows differently and figuring out how to extract only matching ones into that sync job.

**Comment by ofek** [comment]
I don't have time to experiment with upgrading this action this weekend but I plan to manually download like this https://github.com/WordPress/openverse/blob/00a26a65e93fe8ea2785d4a26c3ebce2efc935db/.github/actions/load-img/action.yml#L26-L35

However, rather than an enumeration I will perform pattern matching after I figure out the APIs.

**Comment by webknjaz** [comment]
> I don't have time to experiment with upgrading this action this weekend but I plan to manually download like this https://github.com/WordPress/openverse/blob/00a26a65e93fe8ea2785d4a26c3ebce2efc935db/.github/actions/load-img/action.yml#L26-L35
> 
> However, rather than an enumeration I will perform pattern matching after I figure out the APIs.

I also have some hacking ideas with random names, pattern matching and CLI. Working on a composite action based experiment...

**Comment by TWiStErRob** [comment]
There are mentions of separate upgrades of v3/v4 here breaking and I've had problems in many of my repositories. If anyone is using Renovate and has incompatibility problems, this rule can fix the problem easily (I [added it to my user level Renovate config](https://github.com/TWiStErRob/renovate-config/commit/62f68db39a6edf4e9a21ef629ad5264b08a09dbc) to affect all repos at once.):
```json
"packageRules": [
	{
		"groupName": "actions/upload-artifact and actions/download-artifact",
		"description": "Artifact management is coupled in GitHub Actions. https://github.blog/changelog/2023-12-14-github-actions-artifacts-v4-is-now-generally-available/",
		"matchManagers": ["github-actions"],
		"matchPackageNames": [
			"actions/upload-artifact",
			"actions/download-artifact"
		],
		"matchUpdateTypes": ["major"]
	}
]
```
Note: this above is really explicit, it could work without matching updateTypes and managers too, if you want to generalize.

**Comment by per1234** [comment]
>your example features a linear workflow with a known static artifact names

@webknjaz as indicated by the quote in that comment, my example is specifically in response to @henryiii's comment implying that the **actions/download-artifact** is unable to download multiple artifacts to the same path:

> If there was some way for download-artifact to extract multiple artifacts into a single directory\

The sole purpose of my example was to supplement my comment that I have found that to be untrue. The **actions/download-artifact** is able to download multiple artifacts.

You seem to somehow be interpreting my comment as being intended to dispute the impact of the change of only allowing one upload per artifact. It was not.

>Note all other people who commented here, combined maintain hundreds of workflows too. What unites us is that we all come from the Python ecosystem

That is false. I also maintain hundreds of workflows that were broken by this change and these workflows have nothing to do with Python. It is counterproductive to imply that this impact is specific to Python. The impact is much more widespread than that.

**Comment by webknjaz** [comment]
@per1234 I apologize! I meant to augment the context, not dismiss your input. I did indeed misinterpret your comment, though, and felt like it turns the discussion away from the most problematic part of the breaking change. I didn't mean to imply that this problem is only present in our ecosystem, just wanted to give the context of who all those other people besides you.

**Comment by webknjaz** [comment]
@ofek @henryiii @hynek @woodruffw so this is the UX I'm shooting for: https://github.com/webknjaz/re-actors--download-artifact--test/actions/runs/7242013168/workflow#L38-L42. Feel free to lurk and experiment, but don't rely on the Git identifiers — I'll make my usual versioning setup with immutable tags and rolling branches and will probably force-push the current thing a couple of times later today, before declaring it ready.

**Comment by jnsgruk** [comment]
Not sure how useful or directly applicable this is to OP's problem, but I worked around this using the `gh` cli tool and some reasonably simple `bash` logic here: https://github.com/snapcrafters/ci/pull/24

**Comment by andrewakim** [comment]
👋 Hey everyone!

We appreciate all the feedback you’ve shared so far about upload/download artifact@v4. We’re excited about this release as it includes many improvements, like considerably faster upload/download speeds and the immediate availability of uploaded artifacts. In order to achieve these performance gains, we decided to fundamentally rewrite Artifacts from the ground up, which we’ll share more about in a future blog post.

There was some concern expressed about the immutability of an artifact once uploaded. Previously, each file within an Artifact was individually uploaded to a proxy service before landing in blob storage. This approach allowed actions/upload-artifact to be called on the same object multiple times but it also adversely impacted performance and risked content corruption during concurrent uploads to the same artifact. Artifacts in `v4` are individually sealed archives (zips) that are assembled on the runner and then uploaded directly to blob storage. We believed this change was worth the tradeoff for significant performance gains and to protect the integrity of uploaded artifacts.

We'll be making an update to `actions/download-artifact` by end of day which will include a solution similar to others proposed in this issue:

* The `download-artifact` action will have a new `pattern:` input that will filter the downloaded Artifacts. If you have multiple Artifacts being uploaded in a matrix scenario, you can suffix with other identifiers e.g. `my-artifact-${{ matrix.runs-on }}`, and downloaded them all at once with as `pattern: my-artifact-*`
* In addition, we’ll have another `merge-multiple` boolean input that will extract all the Artifacts’ content to the same destination directory.
```yaml
 - name: Download Matching Artifacts
     uses: actions/download-artifact@v4
     with:
       pattern: my-artifact-*
       path: my-artifact
       merge-multiple: true
 - run: ls -R my-artifact
   # file-macos-latest.txt
   # file-ubuntu-latest.txt
   # file-windows-latest.txt
```

Thank you for your engagement with Artifacts and your valuable feedback.

**Comment by webknjaz** [comment]
> Not sure how useful or directly applicable this is to OP's problem, but I worked around this in a project I help maintain using the `gh` cli tool and some reasonably simple `bash` logic here: [snapcrafters/ci#24](https://github.com/snapcrafters/ci/pull/24).
> 
> Hopefully useful to some folks 😄

That's basically what I've been attempting in my action experiment — GH CLI + passing `name` to `--pattern` + auto-merging. Been trying to keep the UX as close to the v3 as possible, because adding new arguments means more changes for the end-users, which is rather annoying on the scale.

**Comment by schmidtw** [comment]
@andrewakim Thank you for the update.  Do you have any suggestions/guidance for workarounds for the 10 file limit per workflow?  Or should we be using a different action for the following use cases:

- A matrix workflow fans-out to 60+ jobs.  Each uploads 2 uniquely named artifacts.
- A matrix workflow fans-out to 60+ jobs.  Each uploads 1 uniquely named artifact.  A final aggregation step merges the test results into a single report.

**Comment by TWiStErRob** [comment]
@schmidtw the readme says:

> Each **job in a workflow run** now has a limit of 10 artifacts.

Does a job with `matrix:` strategy count as one "job" for you? It works for me: https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/7231579385

**Comment by robherley** [comment]
👋 The changes @andrewakim stated are now released on the `v4` branch.

You can find the new `inputs` documented in the README:
- https://github.com/actions/download-artifact#inputs

As well as a new migration document outlining this use case:
- https://github.com/actions/download-artifact/blob/main/docs/MIGRATION.md

Thanks again for your feedback!

**Comment by nedbat** [comment]
@robherley I haven't tried it yet, but this looks great.  A link to the migration doc from the readme of this repo will help a lot. The first sign many people will see that indicated a needed migration will be an error from upload-artifact.  The breaking change notice in the readme is helpful, and a link to the migration doc will be even better. Thanks!

**Comment by TWiStErRob** [comment]
Thank you for implementing and releasing this, @robherley; it's been a long-awaited change! It's amazing to see the activity on these actions 🤩.

I've tested the the PR branch before the reviews and it worked well. Please consider reviewing the issues linked in https://github.com/actions/download-artifact/pull/259#issuecomment-1861691243 and updating/closing those threads too if relevant.

**Comment by webknjaz** [comment]
@robherley the example in your PR demonstrates `name: my-artifact-${{ matrix.runs-on }}` as an example for naming different artifacts differently. In some cases, that's not enough to guarantee uniqueness. For example, with several matrixes adding to the same pattern or the use of extra `include:` entries. A solution could be to have that suffix generated randomly (with something like `python3 -c 'import uuid; print(uuid.uuid4())'` or `python3 -c 'import secrets; print(secrets.token_hex(16))'`).
I think the upload action migration doc may need better examples with this, and maybe even some kind of a first-class support through a new input.

**Comment by webknjaz** [comment]
@robherley do you know if common directory layout subpaths are merged seamlessly? I don't see examples or tests for that in the PR.

**Comment by bryanmacfarlane** [comment]
I think we should really get rid of the 10 artifact limit.  ASFAIK, there's no real reason to have it (even though most have less which isn't a reason).

**Comment by schmidtw** [comment]
> @schmidtw the readme says:
> 
> > Each **job in a workflow run** now has a limit of 10 artifacts.
> 
> Does a job with `matrix:` strategy count as one "job" for you? It works for me: https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/7231579385

Thank you for pointing that out.  I read it a couple of times, & for whatever reason my brain replaced "job" with "workflow".  10 artifacts per job is perfectly fine.

**Comment by webknjaz** [comment]
FTR here's the guidance I'm giving in one of the shared maintenance orgs I overlook: https://github.com/orgs/aio-libs/discussions/31#discussioncomment-7891886

**Comment by ericLemanissier** [comment]
I've just been bitten too. It's a bad experience for the user that the breaking changes are not mentioned at all in the release page https://github.com/actions/upload-artifact/releases. The corollary is that the breaking changes are not mentioned either in @dependabot's PRs.

Regarding the fix, using `${{ matrix.whatever }}` in the artifact name is not a very general solution. Some workflows have matrices with several dimensions, and some matrix values are not adapted for an artifact name(they can contain forbidden characters).
I think that a best solution is to use `${{ github.job }}-${{ strategy.job-index }}`

**Comment by nedbat** [comment]
Wow, I didn't realize that the release page didn't mention the breaking change...

@andrewakim Thanks for helping get the updated download functionality in place, but as a bit of feedback: your comment starts with two long paragraphs about how great v4 is, waiting until the third paragraph to address our concerns.  This underscores the general feeling of being marketed a product that doesn't do what we need, rather than having peer-to-peer engineering discussions about how to solve our problems.

**Comment by ofek** [comment]
That's not at all how I interpreted it as the first two paragraphs include important context about the implementation that explain why the change was made. Until then, I did not really know how it worked behind the scenes.

**Comment by ekohl** [comment]
I agree it's useful context that should have gone into the release notes when it went out, instead of everyone being notified by dependabot and seeing things were broken without a clear explanation.

**Comment by webknjaz** [comment]
> I agree it's useful context that should have gone into the release notes when it went out, instead of everyone being notified by dependabot and seeing things were broken without a clear explanation.

There's a more dangerous side of this: some projects wouldn't even be notified in PRs when the download action isn't run unconditionally, but is gated by a release request. It'd be quite unpleasant to get the news months after that upgrade got merged, in the middle of a release process.

*Knowing this*, I searched through the PyPA repos the other day and found an actual thing that would break in this exact manner because the updated workflow that got the update is triggered manually: https://github.com/pypa/pipx/pull/1147#discussion_r1429431139.

**Comment by ofek** [comment]
@nedbat Now that the work has been done, is it fine to close this issue? https://github.com/actions/upload-artifact/issues/472#issuecomment-1861820431

**Comment by webknjaz** [comment]
@ofek I think we still need to have a clear example/understanding of how to make unique artifact names with multiple matrixes and multiple reruns. So maybe, we'd wait for more answers?

`${{ github.job }}` proposed by @ericLemanissier, contains a job id which is essentially a YAML key. It's unique within one workflow, but is the same when reruns happen. `${{ strategy.job-index }}` is a job number within a matrix, where it's unique, but job indeces will have the same numbers on rerun. So if a previous (even failing) job succeeded uploading an artifact, its rerun will just error out with no fix w/o re-triggering the entire thing with a separate commit or something.

**Comment by ericLemanissier** [comment]
the best is probably to upload artifacts with `name: prefix-${{ github.run_attempt }}-${{ github.job }}-${{ strategy.job-index }}` and download with `pattern: prefix-${{ github.run_attempt }}-*`

**Comment by webknjaz** [comment]
> and download with `pattern: prefix-${{ github.run_attempt }}-*`

Wouldn't different jobs have different `run_attempt`s if I only rerun a few jobs, but not all?

**Comment by webknjaz** [comment]
> > and download with `pattern: prefix-${{ github.run_attempt }}-*`
> 
> Wouldn't different jobs have different `run_attempt`s if I only rerun a few jobs, but not all?

Perhaps, we need timestamps at the beginning of the pattern, to ensure that the downloading and merging happens in the predictable manner. Though, it's not clearly documented (or tested) if files in newer artifacts would override the older ones.

**Comment by ericLemanissier** [comment]
Actually, ignore my last comment. re-running a job seems to delete artifacts uploaded by said job, because it is able to re-upload an artifact with the same name. In the general case (if you job name has no forbidden characters), the solution is to upload artifacts with name: `prefix-${{ github.job }}-${{ strategy.job-index }}` and download with `pattern: prefix-*`

**Comment by anthony-c-martin** [comment]
> Actually, ignore my last comment. re-running a job seems to delete artifacts uploaded by said job, because it is able to re-upload an artifact with the same name. In the general case (if you job name has no forbidden characters), the solution is to upload artifacts with name: `prefix-${{ github.job }}-${{ strategy.job-index }}` and download with `pattern: prefix-*`

This feels like a regression rather than an intentional breaking change, and it doesn't feel like it should be on the user to have to work around it by adding uniqueness.

In our case, the latest run doesn't have any uploaded artifacts:
https://github.com/Azure/bicep/actions/runs/7252241798/attempts/4

Whereas an older run does:
https://github.com/Azure/bicep/actions/runs/7252241798/attempts/3

The GitHub UI is also busted - for the older run (see my address bar):
<img width="1013" alt="Screenshot 2023-12-19 at 4 23 26 PM" src="https://github.com/actions/upload-artifact/assets/38542602/87b49735-72b5-4bc5-ab1e-4a52d04df8f1">
Yet there are artifacts:
<img width="1247" alt="Screenshot 2023-12-19 at 4 23 45 PM" src="https://github.com/actions/upload-artifact/assets/38542602/5ff1ec8c-d374-4287-9745-30c3b0dcbba4">

For the newer run, a number is displayed:
<img width="1005" alt="Screenshot 2023-12-19 at 4 23 54 PM" src="https://github.com/actions/upload-artifact/assets/38542602/f6db4112-3782-4f7e-9a06-5b6560e0436e">
Yet no artifacts are displayed.

**Comment by shinebayar-g** [comment]
Since matrix jobs don't support output (lol), I have been using an artifact as a shared store for matrix job outputs. Single artifact with different files inside of it. Now this workaround has been broken by this.
Time to find a workaround for my workaround I guess..

**Comment by ofek** [comment]
Matrix jobs certainly do support outputs, I do it here: https://github.com/pypa/hatch/blob/hatch-v1.9.0/.github/workflows/build-hatch.yml#L84-L85

**Comment by shinebayar-g** [comment]
> Matrix jobs certainly do support outputs, I do it here: https://github.com/pypa/hatch/blob/hatch-v1.9.0/.github/workflows/build-hatch.yml#L84-L85

But is that matrix output dynamic?

**Comment by webknjaz** [comment]
> Actually, ignore my last comment. re-running a job seems to delete artifacts uploaded by said job, because it is able to re-upload an artifact with the same name. In the general case (if you job name has no forbidden characters), the solution is to upload artifacts with name: `prefix-${{ github.job }}-${{ strategy.job-index }}` and download with `pattern: prefix-*`

Thank you for checking :pray:

**Comment by webknjaz** [comment]
> Matrix jobs certainly do support outputs, I do it here: https://github.com/pypa/hatch/blob/hatch-v1.9.0/.github/workflows/build-hatch.yml#L84-L85

@ofek all individual matrix jobs likely have the same version in your case, but if each job were to output different values, this value would be overwritten by the last finishing job in the matrix. That's the main problem here.


> But is that matrix output dynamic?
> 
> Edit: I'll add _dynamic_ for clarity.

@shinebayar-g I was googling wrt the current issue the other day and stumbled upon an issue or PR that implements templating for the output keys. I can't find it again now, but the docs confirm that it's possible:
> You can use the `matrix` context to create unique output names for each job configuration.

_(https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs#overview)_

This means that you can have
```yaml
    outputs:
      matrix-output-${{ matrix.thing }}: ${{ steps.matrix-step.outputs.variable }}
```

This likely means that you'd have to inspect the `needs.matrix-job.outputs` context dynamically later on, or reuse your matrix definition somehow, to access said outputs.

**Comment by nedbat** [comment]
I'm a bit lost on how dynamic matrices work and how they affect this because I don't use them.

My matrices are working well using the guidance in the migrating documentation: https://github.com/nedbat/coveragepy/commit/9bda95de6d954f1df7c6dfe50a217c2bb0cd4ab2

**Comment by henryiii** [comment]
I think "dynamic" means different output from each job above. There's also truly dynamic matrices that are generated from a step, like in https://iscinumpy.dev/post/cibuildwheel-2-10-0/#only.

I'm updating the [scientific-python development guide](https://learn.scientific-python.org/development/) to use the suggestions here, and `sp-repo-review` will soon be able to detect and flag repos with workflows that need updating for v4. Those should go live in a day or two if all goes well. We have just merged updated documentation for `cibuildwheel`.

Thanks for providing a good path forward!

**Comment by nedbat** [comment]
> Do we have even a single person working from github working on upload-artifact engaged with this?

@andrewakim and @robherley are marked as "Staff", and have made changes.  There are many comments here, so it's easy to get lost in them, but it's not like nothing is happening.

**Comment by ssbarnea** [comment]
Thanks for pointing, yep I tried to spot some and fail among the rest of the comments. If something is baking is good. We will tame dependabot until we have a fix. I also had to do few rollbacks because I did not realise that v4 was not really working... with coverage.

**Comment by nulano** [comment]
> There are many comments here, so it's easy to get lost in them, but it's not like nothing is happening.

A link to the two comments in this thread if anyone is having a hard time finding them: https://github.com/actions/upload-artifact/issues/472#issuecomment-1861571655, https://github.com/actions/upload-artifact/issues/472#issuecomment-1861820431

**Comment by tonhuisman** [comment]
It's rather simple:
- Changes are not backward compatible
- Changes offer no visible benefit over v3 implementation, AFAICS (that may be my blind spot though)
- Breaks many/all (matrix) builds (using Arduino/C++ here, not Python)
- No migration documentation

Conclusion: Must be a wrong/poor implementation, so no reason whatsoever to implement, and this change should be reverted.

NB: Exactly the same feelings for the `actions/download-artifact` v4 release

**Comment by nedbat** [comment]
@tonhuisman did you miss that they added the migration documentation as requested? Did you overlook the explanation of the benefits?

**Comment by nulano** [comment]
> Changes offer no visible benefit over v3 implementation, AFAICS (that may be my blind spot though)

I have seen some reports that v4 is *significantly* faster in some cases.
Also, the uploaded artifacts are available immediately rather than after all jobs have finished, which is quite a significant improvement IMO.
They've also mentioned that these changes will enable them to finally implement long requested UI improvements; we'll have to wait an see on this one.

Braking the merging capability is unfortunate, but at least they've provided an OK workaround for merging while downloading with download-artifact.
I would like to also see an option for downloading multiple artifacts at once from the summary page, hopefully this will be added at some point.

**Comment by chinoblaize** [comment]
Ps.
Sos

**Comment by tonhuisman** [comment]
> @tonhuisman did you miss that they added the migration documentation as requested? Did you overlook the explanation of the benefits?

@nedbat Thanks, the migration documentation, though somewhat hidden, I have found by now (and implemented).

About the benefits:
- We're using a matrix build that produces ~160 artifacts, with usually 2 files per artifact (we _are_ working on reducing the matrix size but for other reasons, these are PlatformIO ESP builds for a couple (8) of ESP types and several HW configurations). Faster uploads don't really attribute much in reducing the build time here, as that's about 1% of the build time...
- When downloading the matrix-artifacts (for/by combining related artifacts into 8 archives for easier distribution) using `actions/download-artifact@v4` I often see connection issues, breaking the build... 😞 I've switched to trying to assemble into a single archive, and still have the separate artifacts available, but that doesn't solve the download-issues, of course.

About my expectations:
- Improving upload speed should be achievable and still be backward compatible. Backward compatibility is king! (Yes, I've been in software development 'for some time', like 35+ years 😃, it's **important**), though I've seen many things break over the last couple of years, since we started using GH Actions (but also in PlatformIO)

**Comment by ofek** [comment]
> Improving upload speed should be achievable and still be backward compatible. Backward compatibility is king!

GitHub actions use semantic versioning and require an explicit tag or Git reference so introducing a breaking change in a major release is totally standard since you are the one controlling the version. Actions should be treated just like dependencies, there is no difference.

**Comment by tonhuisman** [comment]
> since you are the one controlling the version

Yes, but it would keep me stuck on that older version until they are no longer usable (Node16 vs Node20? once Node16 goes out of support...) or I'm ready to upgrade my scripts to be compatible again, requiring quite some work (mostly testing, but a lot of code-tinkering first)

**Comment by ofek** [comment]
What you speak of is no different than a major version upgrade of a dependency in any ecosystem. Are you then advocating that there should be no breaking changes even between major versions?

**Comment by ericLemanissier** [comment]
I've said it earlier, but I'll repeat it because nothing happened:
As for all major version upgrade with breaking changes, the breaking changes have to be mentioned in the release itself, along with a link to the migration guide. I cannot see either in https://github.com/actions/upload-artifact/releases/tag/v4.0.0

**Comment by bryanmacfarlane** [comment]
@ericLemanissier 💯 on your point.  While there was technically a major version** here what was obviously missed here was "guidance".  Both in the form if "if you used to do x, now do y" and also clearly communicating the breaking changes in the release notes themselves.  The fact that dependabot created PRs doesn't help at all here.  We could have also done a better -preview release with garnering feedback from customers looking for the speed / reliability improvements.  Some scenarios were obviously missed.  The team is discussing all of these, working hard to close the gaps and will do better going forward.  

Thank you so much for the feedback. 

** technical note of compatibility discussed [in docs](https://docs.github.com/en/actions/creating-actions/about-custom-actions#good-practices-for-release-management) - however no excuse for problems above ☝

**Comment by ofek** [comment]
@bryanmacfarlane As an action item you should probably do what others have mentioned and add the [migration guide](https://github.com/actions/download-artifact/blob/main/docs/MIGRATION.md) to the text of the [v4 release](https://github.com/actions/upload-artifact/releases/tag/v4.0.0).

**Comment by bryanmacfarlane** [comment]
@ofek 💯 discussed that with the team.

**Comment by tonhuisman** [comment]
@bryanmacfarlane Thank you for your extensive explanation.

> we would hope that 10x++ perf gains, reliability, and the artifact being immediately available would be a reason to move over and update your yaml (with migration guidance).

Well, possibly the fact we have ca. 160 artifacts from a matrix build (not really sure if that's _many_, but most projects have a lot less), and trying to `action/download-artifact@v4` them in a single step may be the issue leading to connection failures, or maybe just a GH service that needs to be scaled up a bit (it does sometimes succeed), it's the having to re-engineer something that was working OK that makes all this itch a bit.
We're trying to keep stuff up to date, but maybe an alternative could be to go for a Node20 compatible release of the v3 actions?

**Comment by bryanmacfarlane** [comment]
@tonhuisman in the mean time, `v3` is still 100% compatible and works as is before v4 was added cumulatively.  service side is still intact 100% unchanged for that flow as well.  

The team should understand your scenario better.  Maybe create another issue with the scenario and what you're experiencing?  Would probably be good to keep this issue focused on the guidance.  `v4` should theoretically use less connections (and  be more reliable) since it does it per artifact where `v3` went through many requests and multiple hops per file in an artifact (something like node/js artifacts with tons of small files being the worst).  But perhaps there's a problematic scenario?

@robherley

**Comment by webknjaz** [comment]
@bryanmacfarlane FYI what @tonhuisman describes sounds like this old issue I reported a while back that is now more likely to become visible than before: https://github.com/actions/toolkit/issues/1235

**Comment by tonhuisman** [comment]
> @bryanmacfarlane FYI what @tonhuisman describes sounds like this old issue I reported a while back that is now more likely to become visible than before: [actions/toolkit#1235](https://github.com/actions/toolkit/issues/1235)

I've been able to work around that 'download failing' issue by using this [wretry.action](https://github.com/Wandalen/wretry.action) action for now.

**Comment by kvanbere** [comment]
It would be nice if we could maybe upload artifacts to folders or group them visually? I don’t mind uploading all the artifacts as separate zips… yes we also have like 100 matrix builds… but please then add a way to group them rather than splat them all in the same list, and a “Download All” button to avoid a useless GH action to download them all and reupload them in the same zip file.

**Comment by jsoref** [comment]
@bryanmacfarlane I'm going to reply to https://github.com/actions/upload-artifact/issues/472#issuecomment-1870582680
> @tonhuisman We will maintain v3 along with backward compatibility. IF it's ever sunsetted it will go through the long deprecation / announce cycle. v3 should get the proper support including node end of lifes etc. as long as it's supported.

That did *not* happen for any practical interpretation. It's being killed within a month of your comment. That is *not* a lifecycle. The software you're asking us to use still has unreleased [bug fixes](https://github.com/actions/download-artifact/pull/275) that mean it doesn't work properly. The README at the root of these repositories starts with Improvements. It should *not* it *should* start with `Breaking Changes`. It also shouldn't favor ` self hosted runners [sic]` it should favor things that are more commonly used, and that's *clearly* artifact name collisions (and I'm *sure* your team knows that).

There is a single sentence:
> For assistance with breaking changes, see [MIGRATION.md](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md).

But, it's *really* buried and it isn't called out. You could even use the new GitHub markdown syntax to call it out.

Beyond that, this thing isn't supported by GHES which means that if an action author (that's me) actually needed to support GHES customers as well as people using FPT then we'd be screwed. That's really mean. (I don't know that my action is/isn't used by GHES customers, but it's definitely used by enterprises.)

Your teams messed up repeatedly and made it worse here. Fwiw, when my team messes up, we create a new major version that backs out our thing and go back to the drawing board and try to be much more careful about our next rollout. Sure it means that we have much higher major versions in some of our modules, but versions are relatively free. Browser vendors are well over 100 and it turns out that users don't mind.

The release notes for [v4.3.0](https://github.com/actions/upload-artifact/releases/tag/v4.3.0) do not remind people that v4 is incompatible with v3. I think at this point the first line of *every* v4.x release should be "v4.* is incompatible with v3, please see [MIGRATION.md](https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md). -- As noted in https://github.com/actions/upload-artifact/issues/472#issuecomment-1870588031

`actions/*` is using `@v (n+1)` releases for two purposes:
1. when the actions runtime changes
2. when an action's api surface changes

GitHub actions teams _apparently_ committed not to change the actions runtime within a major version, so much so that [github/codeql-action](https://github.com/github/codeql-action/) messed up and merged a node version update to v2 and then undid it.

If you're committed not to change a node runtime within a version, then you should *not* have done an api break in the action.yml surface using `n+1`, you should have used `floor(n/10)+11` instead.

That would have resulted in this release being actions/upload-artifact@v11 and actions/download-artifact@v11 and given you space to release an actions/upload-artifact@v4 and actions/download-artifact@v4 for the node runtime change.

> The model is so different that we even discussed whether it makes sense to create another action but that has another downside of the market place getting fractured.

Yes, that would have been the correct choice. You could easily have created a proper deprecation system where the old action slowed down/warned over time while still allowing it to get runtime updates.

It would have also allowed people to write a more useful migration tool to actually try to migrate users instead of letting dependabot make a mess of things by making PRs that did not work.

**Comment by tonhuisman** [comment]
> > @bryanmacfarlane FYI what @tonhuisman describes sounds like this old issue I reported a while back that is now more likely to become visible than before: [actions/toolkit#1235](https://github.com/actions/toolkit/issues/1235)
> 
> I've been able to work around that 'download failing' issue by using this [wretry.action](https://github.com/Wandalen/wretry.action) action for now.

Unfortunately, that action also has a Node 16 dependency, so now gets a deprecated notification too 😞 
I'm removing that, hoping (praying?) that previous bulk-artifact download issues have actually been resolved... 🙏

**Comment by jsoref** [comment]
Fwiw, for my action, I'm moving to use `gh run download` instead of `actions/download-artifact@v4`.

A feature of `actions/upload-artifact@v4` is that the uploads are available as soon as they're made which means that `gh run download` can retrieve them within a workflow run whereas that wasn't possible for `actions/upload-artifact@v3` things. (From my perspective, `actions/download-artifact@v3` appeared to be using a private api to look under the hood and access artifacts before they were generally available.)

**Comment by tonhuisman** [comment]
Using the plain `actions/download-artifact@v4` now, I'm seeing this runtime warning:
```
(node:1805) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
```
I guess there is still some work to do for the `download-artifact` action 🤔 

On the positive side, it's all downloading fine 👍

**Comment by jsoref** [comment]
Yes, that's tracked as https://github.com/actions/download-artifact/issues/283

**Comment by burnettk** [comment]
i was following the https://github.com/actions/download-artifact/blob/main/docs/MIGRATION.md, which talks about setting the artifact name to something like `my-artifact-${{ matrix.runs-on }}`. this won't actually work, right? this results in artifact names like "my-artifact-" for me. did it intend to say matrix.session or something else?

**Comment by burnettk** [comment]
ok, fair enough, thanks.

**Comment by henryiii** [comment]
That works because the matrix in the example _has_ a `runs-on` key. You need to set it to _your_ key(s), or you can use the less-descriptive but universal `strategy.job-index`, which works for any matrix but just uses numbers.

**Comment by henryiii** [comment]
(https://learn.scientific-python.org/development/guides/repo-review/ will even suggest this if you put in a repo that would fail with v4)

**Comment by jsoref** [comment]
https://github.com/actions/upload-artifact/issues/472#issuecomment-1924569423's feedback is valid, the text:

Conceptually, here's a change to (hopefully) improve that experience:
```diff
 In v4, Artifacts are immutable (unless deleted).
-So you must change each of the uploaded Artifacts to have a different name
+So you give each uploaded Artifact a distinct name
+(in this example the matrix is has a `.runs-on` property which is used to provide a distinct name,
+but how you choose a distinct name will vary)
 and filter the downloads by name to achieve the same effect:
```

**Comment by nowakweronika** [comment]
In v4, Artifacts are immutable. So you must change each of the uploaded Artifacts to have a different name:
```diff
name: Upload unit tests results
  if: always()
- uses: actions/upload-artifact@v3
+ uses: actions/upload-artifact@v4
  with:
-     name: test-results
+     name: test-results-unit
          path: |
          */build/test-results/unit
name: Upload instrumentation tests results
  if: always()
- uses: actions/upload-artifact@v3
+ uses: actions/upload-artifact@v4
  with:
-     name: test-results
+     name: test-results-instrumentation
          path: |
          */build/test-results/instrumentation
```
And then you can merge it:
```
merge:
  runs-on: ubuntu-latest
  needs:
    - android-tests
    - unit-tests
  steps:
    - name: Merge Artifacts
      uses: actions/upload-artifact/merge@v4
      with:
        name: test-results
        pattern: test-results-*
```
You can see full PR here: https://github.com/appunite/Loudius/pull/190/files

**Comment by rcdailey** [comment]
Version 4 has been ultimately a downgrade for me, as it removes the ability to replace archives. I have a job that runs in my pipeline for only certain architectures that does apple code signing. In this case, it downloads a artifact (which contains a binary), signs it, and then re-uploads the artifact, replacing the original. But I cannot do this with v4. The merge action does not address this workflow either.

My artifacts are named after architectures in a matrix build; for example: `linux-x64`, `win-x64`, `osx-x64`. It's important that the names remain the same, even if they get replaced, because my final release job iterates the artifacts and attaches them to a Github Release. So it needs to be able to find `osx-x64`, not something like `osx-x64-signed`, because I won't have equivalent versions for linux, for example (e.g. `linux-x64-signed` doesn't make sense). The name the artifacts start with is the name they have to end with.

Immutable packages don't need to _change_, but we should have the ability to _replace them_.

**Comment by rcdailey** [comment]
> > The merge action does not address this use case either.
> 
> I just successfully upgraded to v4 a few days ago where I do the exact same thing that you're talking about... [pypa/hatch@03b1bca#diff-c4330c2cfbadba3e17aef4105738e5644a48e08e86096a32e2a6625fcd83b957](https://github.com/pypa/hatch/commit/03b1bca8dcf0cbe6d43d9c014d5df47bcde036d6#diff-c4330c2cfbadba3e17aef4105738e5644a48e08e86096a32e2a6625fcd83b957)

Your workflow file appears to be structured differently than mine. Here is how I'm doing things:
https://github.com/recyclarr/recyclarr/blob/master/.github/workflows/build.yml

Taking a peek at your workflow, it looks like you make `staged-` artifacts for every platform, including macos. However in my case, I do not have additional preparation work to do, so technically only macos needs staging artifacts. Upgrading to v4 shouldn't force me to create unnecessary staging/temporary artifacts when in v3 a simple "reupload" was sufficient for the special cases.

**Comment by jsoref** [comment]
@rcdailey there's now a `overwrite: true` option...

**Comment by rcdailey** [comment]
> @rcdailey there's now a `overwrite: true` option...

How embarrassing; I completely missed this in the documentation. Thank you very much for pointing it out. I tested with this setting and it works great!

**Comment by pdurbin** [comment]
![383340804-031aa6d3-6b54-4735-9082-052e9813c44d](https://github.com/user-attachments/assets/f251601d-ee2e-4b2c-aba3-8348e28ca137)
